### ISSUE TITLE
[2.0~beta4/cmake] is installation path of FindOpenCL.cmake incorrect?

### ISSUE THREAD

**Issue by cdluminate** [oneapi-src/oneDNN#650]
Shouldn't  `${LIB_CONFIG_INSTALL_DIR_COMPAT}` be corrected into `${LIB_CONFIG_INSTALL_DIR}`?

https://github.com/intel/mkl-dnn/blob/v2.0-beta04/src/CMakeLists.txt#L209

**Comment by rsdubtso** [comment]
Hi @cdluminate. Thanks for the reported issue.

I may be missing something, but I think the current code is correct.

The first `install()`, https://github.com/intel/mkl-dnn/blob/53b9208b9da5639a60bfdbff1dd3ab24e84388be/src/CMakeLists.txt#L205-L207
copies `FindOpenCL.cmake` to the cmake package install location.

The second `install()`,
https://github.com/intel/mkl-dnn/blob/53b9208b9da5639a60bfdbff1dd3ab24e84388be/src/CMakeLists.txt#L208-L209
copies the `FindOpenCL.cmake` to a location where it would be found by projects that still rely on the library being called MKL-DNN, hence the 'COMPAT' in the name.

**Comment by cdluminate** [comment]
Thank you for the quick response! Actually I came across a installation failure in a clean chariot. I will continue to track the source of problem and get back to you later.

**Comment by cdluminate** [comment]
```

**Comment by cdluminate** [comment]
These are my build options
```makefile
# https://intel.github.io/mkl-dnn/dev_guide_build_options.html
override_dh_auto_configure:
        dh_auto_configure -- \
                -DCMAKE_BUILD_TYPE=RelWithDebInfo \
                -DMKLDNN_ARCH_OPT_FLAGS="" \
                -DDNNL_GPU_RUNTIME=OCL \
                -DDNNL_INSTALL_MODE=DEFAULT
```

**Comment by rsdubtso** [comment]
Hi @cdluminate, sorry for a delay. Can you please post more details on how to reproduce this in e.g. a docker container? I've tried pulling `debian:latest` (`/etc/debian_version` says that it is a 10.2), building and installing the library using the CMake options above, but did not observe any issues.

**Comment by rsdubtso** [comment]
Just for reference:
```
Install the project...
-- Install configuration: "RelWithDebInfo"
-- Installing: /root/mkl-dnn/install/share/doc/dnnl/LICENSE
-- Installing: /root/mkl-dnn/install/share/doc/dnnl/README
-- Installing: /root/mkl-dnn/install/lib/libdnnl.so.1.91
-- Installing: /root/mkl-dnn/install/lib/libdnnl.so.1
-- Installing: /root/mkl-dnn/install/lib/libdnnl.so
-- Installing: /root/mkl-dnn/install/include/dnnl_config.h
-- Installing: /root/mkl-dnn/install/include/dnnl_version.h
-- Installing: /root/mkl-dnn/install/include/dnnl.h
-- Installing: /root/mkl-dnn/install/include/dnnl.hpp
-- Installing: /root/mkl-dnn/install/include/dnnl_debug.h
-- Installing: /root/mkl-dnn/install/include/dnnl_types.h
-- Installing: /root/mkl-dnn/install/include/mkldnn.h
-- Installing: /root/mkl-dnn/install/include/mkldnn.hpp
-- Installing: /root/mkl-dnn/install/include/mkldnn_config.h
-- Installing: /root/mkl-dnn/install/include/mkldnn_debug.h
-- Installing: /root/mkl-dnn/install/include/mkldnn_dnnl_mangling.h
-- Installing: /root/mkl-dnn/install/include/mkldnn_types.h
-- Installing: /root/mkl-dnn/install/include/mkldnn_version.h
-- Installing: /root/mkl-dnn/install/lib/cmake/dnnl/dnnl-config.cmake
-- Installing: /root/mkl-dnn/install/lib/cmake/dnnl/dnnl-config-version.cmake
-- Installing: /root/mkl-dnn/install/lib/cmake/dnnl/dnnl-targets.cmake
-- Installing: /root/mkl-dnn/install/lib/cmake/dnnl/dnnl-targets-relwithdebinfo.cmake
-- Installing: /root/mkl-dnn/install/lib/libmkldnn.so
-- Installing: /root/mkl-dnn/install/lib/libmkldnn.so.1
-- Installing: /root/mkl-dnn/install/lib/libmkldnn.so.1.91
-- Installing: /root/mkl-dnn/install/lib/cmake/dnnl/FindOpenCL.cmake
-- Installing: /root/mkl-dnn/install/lib/cmake/mkldnn/FindOpenCL.cmake

root@a5d33aabd20c:~/mkl-dnn/build# ls /root/mkl-dnn/install/lib/cmake/mkldnn/
FindOpenCL.cmake  mkldnn-config-version.cmake  mkldnn-config.cmake  mkldnn-targets-relwithdebinfo.cmake  mkldnn-targets.cmake

```

**Comment by rsdubtso** [comment]
Closing as cannot reproduce. @cdluminate : feel free to reopen with more info.

**Comment by kurnevsky** [comment]
Just had a similar problem trying to build the version 1.4 with `-DDNNL_GPU_RUNTIME=OCL` on arch:
```
CMake Error at /home/kurnevsky/workspace/onednn/src/oneDNN-1.4/cmake/gen_mkldnn_compat_cmakes.cmake:25 (file):
  file problem creating directory: /usr/lib/cmake/mkldnn


-- Installing: /home/kurnevsky/workspace/onednn/pkg/onednn/usr/lib/libmkldnn.so
-- Installing: /home/kurnevsky/workspace/onednn/pkg/onednn/usr/lib/libmkldnn.so.1
-- Installing: /home/kurnevsky/workspace/onednn/pkg/onednn/usr/lib/libmkldnn.so.1.4
-- Installing: /home/kurnevsky/workspace/onednn/pkg/onednn/usr/lib/cmake/dnnl/FindOpenCL.cmake
CMake Error at src/cmake_install.cmake:192 (file):
  file INSTALL cannot find "/usr/lib/cmake/dnnl/FindOpenCL.cmake": No such
  file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:46 (include)
```

**Comment by rsdubtso** [comment]
Hm... this is interesting. Maybe there's some nondermenism? @kurnevsky does pkgbuild add any parallelization options to `MAKEFLAGS`?

**Comment by kurnevsky** [comment]
Yes, if it's configured in `/etc/makepkg.conf` and I set to `-j4`. Will try to build with one thread.

**Comment by kurnevsky** [comment]
The same error when building without `-j4`...

**Comment by rsdubtso** [comment]
Can you reproduce this without `pkgbuild`? Can you please run `cd src && cmake -P cmake_install.cmake` from the build directory? And can you please the contents of `build/src/cmake_install.cmake`?

**Comment by kurnevsky** [comment]
> Can you reproduce this without pkgbuild?

Yes, with following:

```sh
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX="/usr" -DDNNL_GPU_RUNTIME=OCL
mkdir dest
DESTDIR=./dest make install
```

> Can you please run cd src && cmake -P cmake_install.cmake from the build directory?

```
-- Install configuration: "Release"
-- Up-to-date: /usr/share/doc/dnnl/reference
CMake Error at cmake_install.cmake:41 (file):
  file INSTALL cannot set permissions on "/usr/share/doc/dnnl/reference":
  Operation not permitted.
```

> And can you please the contents of build/src/cmake_install.cmake?

[cmake_install.cmake.txt](https://github.com/oneapi-src/oneDNN/files/4501750/cmake_install.cmake.txt)

**Comment by rsdubtso** [comment]
> ` Operation not permitted.`
This is due to insufficient permissions. Can you change the `CMAKE_INSTALL_PREFIX` to some path that your user can write to?

> cmake_install.cmake.txt
This is the top-level `build/cmake_install.cmake`. I need the one from `build/src'.

I started suspecting that this issue is specific to packaging situations because something intercepts the `install` but does not intercept the `cp` (or whatever else).

**Comment by kurnevsky** [comment]
> Can you change the CMAKE_INSTALL_PREFIX to some path that your user can write to?

```
-- Install configuration: "Release"
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/libdnnl.so.1.4
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/libdnnl.so.1
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/libdnnl.so
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/dnnl_config.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/dnnl_version.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/dnnl.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/dnnl.hpp
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/dnnl_debug.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/dnnl_threadpool_iface.hpp
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/dnnl_types.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/mkldnn.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/mkldnn.hpp
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/mkldnn_config.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/mkldnn_debug.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/mkldnn_dnnl_mangling.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/mkldnn_types.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/include/mkldnn_version.h
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/cmake/dnnl/dnnl-config.cmake
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/cmake/dnnl/dnnl-config-version.cmake
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/cmake/dnnl/dnnl-targets.cmake
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/cmake/dnnl/dnnl-targets-release.cmake
CMake Error at /home/kurnevsky/workspace/oneDNN/cmake/gen_mkldnn_compat_cmakes.cmake:25 (file):
  file problem creating directory: /usr/lib/cmake/mkldnn


-- Up-to-date: /usr/lib/libmkldnn.so
-- Up-to-date: /usr/lib/libmkldnn.so.1
-- Up-to-date: /usr/lib/libmkldnn.so.1.4
-- Installing: /home/kurnevsky/workspace/oneDNN/build/src/./tmp/lib/cmake/dnnl/FindOpenCL.cmake
CMake Error at cmake_install.cmake:192 (file):
  file INSTALL cannot find "/usr/lib/cmake/dnnl/FindOpenCL.cmake": No such
  file or directory.
```

> This is the top-level build/cmake_install.cmake. I need the one from build/src.

Sorry, here it is: [cmake_install.cmake.txt](https://github.com/oneapi-src/oneDNN/files/4505287/cmake_install.cmake.txt)

**Comment by rsdubtso** [comment]
> `  file problem creating directory: /usr/lib/cmake/mkldnn`

This is still quite weird.
- Is this with a clean rebuild? Can you please share the exact cmake command line?
- Can you please apply the patch from #709 and retest?

Sorry this is taking so long. I just want to be able to fix this once and for all.

**Comment by kurnevsky** [comment]
> Is this with a clean rebuild? Can you please share the exact cmake command line?

Yes, I specified commands in https://github.com/oneapi-src/oneDNN/issues/650#issuecomment-616355834 and after I just executed `cmake -DCMAKE_INSTALL_PREFIX=./tmp -P cmake_install.cmake`.

> Can you please apply the patch from #709 and retest?

It works with this patch, thanks.

> Sorry this is taking so long. I just want to be able to fix this once and for all.

No problem, I'm not in hurry :)

**Comment by rsdubtso** [comment]
Fixed in master in 2855d429247aa743af456dd7f8e03ab5df1a2400. It should be promoted to dev-v2 eventually. Closing.
